prov/util/src/util_mr_cache.c

```diff
@@ -134,6 +134,8 @@ static void util_mr_uncache_entry(struct ofi_mr_cache *cache,
 	if (entry->use_cnt == 0) {
 		dlist_remove(&entry->list_entry);
 		dlist_insert_tail(&entry->list_entry, &cache->deferred_list);
+		cache->deferred_cnt++;
+		cache->deferred_size += entry->info.iov.iov_len;
```

Comment:
Until the entry is flushed, rather than tracking it in another location, it may be cleaner to track the cnt and size as part of the cache.

The uncached_cnt / uncached_size fields are only there for debugging purposes.  Those two fields don't impact the behavior at all, and can basically be ignored.  The cached_cnt / cached_size do.  Rather than introduce new fields, we should be able to adjust how cnt/size are updated, so that they track all registrations being watched by the cache, regardless if the entry is in the flush_list or rbmap.

I would remove cnt/size adjustments from  util_mr_uncache_entry_storage(), and look at removing that function.  It's manipulation of cnt/size results in some hidden modifications of those values, compared to the places where it is invoked.  For example, look at util_mr_cache_create().  That function updates cnt/size directly, but ends up doing so in an odd way because util_mr_uncache_entry_storage() updates the values behind the scene.  util_mr_cache_create() should update cnt/size after calling ofi_monitor_subscribe(), but can't because of the error flow.

Even this patch doesn't ever decrement deferred_cnt/size.  So it's off in that respect.  I think we want the cnt/size values to adjust near all calls to util_mr_free_entry().  It can't be done within that call because of locking issues, but should be immediately before the call.  Whether we adjust uncached or cached cnt/size would be based on the context of the caller.  E.g. ofi_mr_cache_flush() should adjust cache_cnt/size.  ofi_mr_cache_delete() already handled uncached, with a proper check.

Reply:
I think you are suggesting to include the flush_size in cached_size, which is a great idea, and can solve the issue.

However, I still think the have a separate counter for `flush_list` has its advantage.

Currently, `ofi_mr_cach_flush` can do two things: clear the `flush_list` and clear the LRU list.

Clearing flush_list does not have any performance issue, but clearing the LRU list does. (because we flush out a MR that can be re-used).

With the counter to `flush_list`, we can prioritize clearing `flush_list`: if `flush_list` is not empty, clear the flush_list. Only when flush_list is empty, we clear the LRU list.

I was also hoping that we can remove the argument `clear_lru` to `ofi_mr_cache_flush`, and split that function into two: `ofi_mr_cache_clear_staled` and `ofi_mr_cache_flush_lru`, making the function name more descriptive.